### ISSUE TITLE
Improved Pooling

### ISSUE THREAD

**Issue by jonathanolson** [phetsims/phet-core#35]
As noted by @samreid in the Area Model review, we have a lot of ugly boilerplate to set up pooling in an efficient manner.

It looks like we can do better in a few different ways (according to my current prototype), so I'll be committing it and iterating so it can be used instead of the current suboptimal pattern.

For instance, the "old" pattern is approximately:
```js
function F( x, y ) {
  this.initialize( x, y );
}
phetCore.inherit( Object, F, {
  initialize: function( x, y ) {
    this.x = x;
    this.y = y;
    return this;
  },
  getSum: function() {
    return this.x + this.y;
  }
} );
phetCore.Poolable.mixInto( F, {
  constructorDuplicateFactory: function( pool ) {
    return function( x, y ) {
      if ( pool.length ) {
        return pool.pop().initialize( x, y );
      }
      else {
        return new F( x, y );
      }
    };
  }
} );
```

which can be replaced with the "new" pattern:
```js
function F( x, y ) {
  this.x = x;
  this.y = y;
}
phetCore.inherit( Object, F, {
  getSum: function() {
    return this.x + this.y;
  }
} );
phetCore.ExperimentalPoolable.mixInto( F );
```

Basically, through scary hacky magic that doesn't appear to slow things down or allocate anything more, we're able to handle all of the pooling tasks for creating/reinitializing through the constructor itself, and you don't have to provide the boilerplate function.

**Comment by jonathanolson** [comment]
Also going to look into ways around https://github.com/phetsims/phet-core/issues/7 so we can create even less garbage.

**Comment by jonathanolson** [comment]
@samreid some time to review current patterns (and any improvements) as part of the Area Model code review would be nice.

**Comment by samreid** [comment]
Generally this looks preferable to the prior implementation of Poolable, nice work!

> scary hacky magic 

I'm not too worried about the `bind.bind`--it seems unlikely to break or be deoptimized in the future.

> [pool is public] so that many items can be added at a time

Perhaps add a specific method for this, instead of requiring clients to use `Array.prototype.push.apply` or equivalent.

> new DefaultConstructor

Can you comment on `new DefaultConstructor`?  I'm not familiar with that parentheses-less form.  A code comment should clarify it.  Or if it is the same as `new DefaultConstructor()`, consider adding the parens so it will be clearer for the team.

> `return new ( partialConstructor.apply( null, arguments ) );`

does that line need the extra parens?  OK with me if you leave them for clarity.

Also, I don't see any usage, examples or tests for ExperimentalPoolable's options.  I'd recommend adding basic sanity unit tests for ExperimentalPoolable to make sure things are working as intended.

Anything else I should look at specifically?

**Comment by jonathanolson** [comment]
> Perhaps add a specific method for this, instead of requiring clients to use Array.prototype.push.apply or equivalent.

I might actually wait until the need comes up. Maybe I should make it private? Thoughts?

> Can you comment on `new DefaultConstructor`?

They look equivalent, I'm going to add the parentheses to be consistent.

> does that line need the extra parens? OK with me if you leave them for clarity.

I left it for clarity, since it is an atypical usage of new.

> Also, I don't see any usage, examples or tests for ExperimentalPoolable's options. I'd recommend adding basic sanity unit tests for ExperimentalPoolable to make sure things are working as intended.

Assigning myself to add unit tests.

> Anything else I should look at specifically?

Can't think of anything, thanks!

**Comment by jonathanolson** [comment]
@pixelzoom I don't see how this would be blocking publication. The "new" pooling setup is only used directly in area-model right now. Over time I'll probably move more pooling usage to the new method.

**Comment by samreid** [comment]
> I might actually wait until the need comes up. Maybe I should make it private? Thoughts?

Let's make it private for now.

**Comment by pixelzoom** [comment]
I was not aware of where pooling might be used, and this seemed like a significant change, so (as I said on Slack) I made my best guess.   In any case, shouldn't this be wrapped up so that others can use it, and since this "prototype code" now appears in several published sims? (the Area Model suite).

**Comment by samreid** [comment]
Are you planning to tackle them as a batch?  If so, one issue seems best.

**Comment by jonathanolson** [comment]
> Are you planning to tackle them as a batch? If so, one issue seems best.

Probably not.

**Comment by pixelzoom** [comment]
Is this new pooling approach completed, or is it still "current prototype"?

**Comment by jonathanolson** [comment]
> Is this new pooling approach completed, or is it still "current prototype"?

It is basically completed, although there's a good chance I'll add an option to override how new-instance construction is handled to support how Matrix3 is pooled.

It's still called ExperimentalPoolable because there were no suggestions on Slack dev-public for a better "temporary" name.

**Comment by pixelzoom** [comment]
Seems like it would be good to (1) wrap this up so that it's no longer experimental/temporary, and (2) chip away at converting "old" style to new.  How you do that is up to you, I don't have a preference since I haven't used old style.

**Comment by jonathanolson** [comment]
There were some concerns brought up in Slack about using the constructor function also for reinitialization (setting up state from an object pulled from the pool), so it would probably be good to discuss this at the developer meeting.

**Comment by samreid** [comment]
What if calling the constructor twice ends up linking a `Property` twice?

**Comment by jonathanolson** [comment]
> What if calling the constructor twice ends up linking a Property twice?

Then it's absolutely a bug, just the same as if initialize() linked to a Property twice.

**Comment by samreid** [comment]
I don't see any usages of `dispose` in Poolable.js.  Does it need something like that (to unlink or clean up) before calling constructor again?

Also, I believe 3rd line in this snippet is the one that runs the constructor again on a live instance.  If that's true, please comment accordingly:

```js
          if ( pool.length ) {
            result = pool.pop();
            initialize.apply( result, arguments );
          }
```

**Comment by jonathanolson** [comment]
> I don't see any usages of dispose in Poolable.js. Does it need something like that (to unlink or clean up) before calling constructor again?

That's been the responsibility of the code that puts it into the pool (to clean out references).

Maybe it would be beneficial (as part of the discussion) for me to show an overview of how pooling is used?

> Also, I believe 3rd line in this snippet is the one that runs the constructor again on a live instance. If that's true, please comment accordingly:

It may not be the constructor itself. How would you recommend I document it?

**Comment by andrealin** [comment]
@jonathanolson will list examples of code that re use the constructor, and ones that do not. And will tag for developer meeting in two weeks.

**Comment by jonathanolson** [comment]
Example with an initialize method:
```js
/**
 * Average pooling example
 * @constructor
 * @mixes Poolable
 *
 * @param {Node} content
 * @param {Property.<number>} massProperty
 * @param {Range} range
 */
function Something( content, massProperty, range ) {
  this.initialize( content, massProperty, range );
}

namespace.register( 'Something', Something );

inherit( Events, Something, {
  /**
   * Initializes Something, implements construction for pooling.
   * @private
   *
   * @param {Node} content
   * @param {Property.<number>} massProperty
   * @param {Range} range
   * @returns {Something} - Returns 'this' reference, for chaining
   */
  initialize: function( content, massProperty, range ) {
    // @private {Node}
    this.content = content;

    // @private {*}
    this.somethingExpensive = this.somethingExpensive || createSomethingExpensive();

    // ...
  },

  /**
   * Releases references, and frees it to the pool.
   * @public
   */
  dispose: function() {
    // does whatever it needs to remove references, unlink, etc.
    this.content = null;

    this.freeToPool();
  }
} );

Poolable.mixInto( Something, {
  initialize: Something.prototype.initialize
} );
```

Example using the constructor:
```js
/**
 * Average pooling example
 * @constructor
 * @mixes Poolable
 *
 * @param {Node} content
 * @param {Property.<number>} massProperty
 * @param {Range} range
 */
function Something( content, massProperty, range ) {
  // @private {Node}
  this.content = content;

  // @private {*}
  this.somethingExpensive = this.somethingExpensive || createSomethingExpensive();

  // ...
}

namespace.register( 'Something', Something );

inherit( Events, Something, {
  /**
   * Releases references, and frees it to the pool.
   * @public
   */
  dispose: function() {
    // does whatever it needs to remove references, unlink, etc.
    this.content = null;

    this.freeToPool();
  }
} );

Poolable.mixInto( Something );
```

**Comment by pixelzoom** [comment]
7/12/18 dev meeting:

• Calling the constructor more than once is (likely) incompatible with ES6 `class`
• Use `initialize` and `dispose` in types that are `Poolable`.
• By convention, `initialize` will have same parameters as constructor, and doc will say "see constructor" instead of duplicating `@param` annotations.  
• For anything `Poolable`, the reader should consult both the constructor and `initialize` for documentation.
• @jonathanolson will back out existing code that is recalling constructor and use new pattern.
